system/security/LdapSecurity/ldapconnection.cpp

```diff
@@ -981,6 +981,202 @@ class CLDAPMessage
     inline operator LDAPMessage *() const { return msg; }
 };
 
+//------------------------------------------------------
+// class CPagedLDAPSearch
+//
+//  Performs a "paged" LDAP search, which allows for
+//  searching when there are over 1000 matching entries. Also
+//  works when there are less than 1000, but not as efficient as
+//  simply calling ldap_search_ext_s()
+//------------------------------------------------------
+#define MAX_ENTRIES 1000
+class CPagedLDAPSearch
+{
+private:
+    LDAP *          m_pLdapConn;
+    char *          m_pszDN;
+    unsigned long   m_scope;
+    char *          m_pszFilter;
+    char * *        m_pszAttrs;
+
+    bool            m_morePages;
+    struct berval * m_pCookie;
+    LDAPMessage *   m_pPageEntry;
+    LDAPMessage *   m_pPageBlock;
+
+    //---------------------------------------------------
+    // Request next page of search results
+    //---------------------------------------------------
+    bool requestNextPage()
+    {
+        if (!m_morePages)
+            return false;
+        LDAPControl * pageControl = NULL;
+        LDAPControl **  returnedCtrls = NULL;
+        try
+        {
+            int rc = ldap_create_page_control(m_pLdapConn, MAX_ENTRIES, m_pCookie, false, &pageControl);//cookie gets set on first call to ldap_parse_page_control()
+            if (rc != LDAP_SUCCESS)
+            {
+                int err = GetLastError();
+                throw MakeStringException(-1, "ldap_create_page_control failed with 0x%x (%s)",err, ldap_err2string( err ));
+            }
+
+            LDAPControl * svrCtrls[] = { pageControl, NULL };
+
+            if (m_pPageBlock)
+                ldap_msgfree(m_pPageBlock);
+            TIMEVAL timeOut = {LDAPTIMEOUT,0};
+            rc = ldap_search_ext_s(m_pLdapConn, m_pszDN, m_scope, m_pszFilter, m_pszAttrs, 0, svrCtrls, NULL, &timeOut, 0, &m_pPageBlock);
+            if (rc != LDAP_SUCCESS)
+            {
+                int err = GetLastError();
+                if (err && rc != LDAP_PARTIAL_RESULTS)//389DirectoryServer sometimes returns rc, but GetLastError returns 0. In this scenario continuing the query succeeds
+                {
+                    ldap_control_free(pageControl);
+                    if (m_pCookie)
+                    {
+                        ber_bvfree(m_pCookie);
+                        m_pCookie = NULL;
+                    }
+                    throw MakeStringException(-1, "ldap_search_ext_s failed with 0x%x (%s)",err, ldap_err2string( err ));
+                }
+            }
+
+            int l_errcode;
+            rc = ldap_parse_result(m_pLdapConn, m_pPageBlock, &l_errcode, NULL, NULL, NULL, &returnedCtrls, false);
+
+            if (m_pCookie)
+            {
+                ber_bvfree(m_pCookie);
+                m_pCookie = NULL;
+            }
+
+            if (rc != LDAP_SUCCESS)
+            {
+                ldap_control_free(pageControl);
```

Comment:
I would have been inclined to create a little wrapper class to contain the pageControl ptr and ensure it was freed on destruction - that way you don't need to check every possible return/throw to make sure it is freed...

But I'm not going to hold up this merge for that suggestion... maybe next time.